### PR TITLE
Perf zerofiers

### PR DESCRIPTION
**Pull request by diegokingston**
Changed the way the boundary zerofiers are calculated. This reduces the amount of zerofiers to be calculated, at the expense of more complicated logic. Results in 10% improvement in proving

First, we filter all the points where the boundary constraints hold and store them inside a vector. We compute the zerofiers just by doing (w - g^p), where w is an element of the coset and we perform batch inversion. Then, we calculate degree adjustment and correct each boundary polynomial by that factor. Afterwards, we go through each column, look at which zerofiers apply and update the boundary polynomial by multiplying it with the corresponding zerofier. Finally, we sum all the evaluations for each column to obtain the contribution of the BC to the composition polynomial.

### CODE REVIEW

File: src/starks/constraints/boundary.rs [hunk 1]
```diff
@@ -59,6 +60,22 @@ impl<F: IsField> BoundaryConstraints<F> {
...
+            .collect()
+    }
+
+    pub fn cols_for_boundary(&self) -> Vec<usize> {
+        self.constraints
+            .iter()
+            .unique_by(|elem| elem.col)
+            .map(|v| v.col)
+            .collect()
+    }
```
Comment (Oppen) [inline src/starks/constraints/boundary.rs:77]: This function seems to be unused.

File: src/starks/constraints/boundary.rs [hunk 2]
```diff
@@ -59,6 +60,22 @@ impl<F: IsField> BoundaryConstraints<F> {
             .collect()
     }
 
+    pub fn steps_for_boundary(&self) -> Vec<usize> {
+        self.constraints
+            .iter()
+            .unique_by(|elem| elem.step)
+            .map(|v| v.step)
+            .collect()
+    }
```
Comment (Oppen) [inline src/starks/constraints/boundary.rs:69]: There are two uses of this and both only iterate once over the result. Consider returning an iterator instead:
```suggestion
    pub fn steps_for_boundary(&self) -> impl Iterator<Item = usize> {
        self.constraints
            .iter()
            .cloned()
            .unique_by(|elem| elem.step)
            .map(|v| v.step)
    }
```

File: src/starks/constraints/evaluator.rs [hunk 3]
```diff
@@ -59,14 +59,40 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
         let n_trace_colums = self.trace_polys.len();
         let boundary_constraints = &self.boundary_constraints;
 
+        let boundary_steps = self.boundary_constraints.steps_for_boundary();
+
+        let boundary_zerofiers_inverse_evaluations: Vec<Vec<FieldElement<F>>> = boundary_steps
+            .iter()
```
Comment (Oppen) [inline src/starks/constraints/evaluator.rs:65]: You can probably remove this now.

File: src/starks/constraints/evaluator.rs [hunk 4]
```diff
@@ -85,30 +111,47 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
...
+                .collect::<Vec<FieldElement<F>>>();
+
+            // Multiply each column for their correspondent zerofiers
+            let mut acc = boundary_polys_evaluations[col].clone();
+            // Step contains the indexes for the boundary constraints
+            // Of the respective column we are interested in
+            for column_step in boundary_constraints.steps(col).iter() {
+                let step_index = boundary_constraints
+                    .steps_for_boundary()
+                    .iter()
```
Comment (Oppen) [inline src/starks/constraints/evaluator.rs:134]: And this.

File: src/starks/constraints/evaluator.rs [hunk 5]
```diff
@@ -85,30 +111,47 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
...
+                    .position(|x| *x == *column_step)
+                    .expect("Should be here");
```
Comment (Oppen) [inline src/starks/constraints/evaluator.rs:136]: Is this logically guaranteed or could this fail on bad input? If it's the latter, we should return an error instead.

File: src/starks/constraints/evaluator.rs [hunk 6]
```diff
@@ -183,7 +223,7 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
                 .collect();
 
         // Iterate over trace and domain and compute transitions
-        for (i, d) in domain.lde_roots_of_unity_coset.iter().enumerate() {
+        for (i, _d) in domain.lde_roots_of_unity_coset.iter().enumerate() {
```
[the remark this answers is not in the document]
Comment (diegokingston) [reply src/starks/constraints/evaluator.rs]: Solved

File: src/starks/constraints/evaluator.rs [hunk 7]
```diff
@@ -85,30 +111,45 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
...
+            let mut acc = boundary_polys_evaluations[col].clone();
+            // Step contains the indexes for the boundary constraints
+            // Of the respective column we are interested in
+            for column_step in boundary_constraints.steps(col).iter() {
+                let step_index = boundary_steps
+                    .iter()
+                    .position(|x| x == column_step)
+                    .expect("Should be here");
+                // (,) ((,),)
+                // acc
```
Comment (MauroToscano) [inline src/starks/constraints/evaluator.rs:137]: This is an unused comment

```suggestion
```

**Review by MauroToscano — APPROVED** [review]

**Review by Oppen — APPROVED** [review]

**Review by gabrielbosio — APPROVED** [review]

**Pull request merged**